### PR TITLE
feat: Implement GSP-700 Config Features and DefaultPairs via Connection String

### PR DESCRIPTION
**Pull request by JinnyYi**
ref: https://github.com/beyondstorage/go-storage/issues/680
ref: https://github.com/beyondstorage/go-storage/issues/704

### CODE REVIEW

**Comment by Xuanwo** [comment]
Please start RFC first before implement this feature.

File: tests/connstr_test.go [hunk 1]
```diff
@@ -84,6 +84,30 @@ func TestFromString(t *testing.T) {
...
+			"tests://abc/tmp/tmp1?default_content_type=application/text&storage_class=sc",
+			[]Pair{
+				pairs.WithName("abc"),
+				pairs.WithWorkDir("/tmp/tmp1"),
+				WithDefaultContentType("application/text"),
+				WithStorageClass("sc"),
+			},
+			nil,
+		},
+		//{
```
Comment (Xuanwo) [inline tests/connstr_test.go:98]: Do we need this test?

File: tests/generated.go [hunk 2]
```diff
@@ -591,8 +681,36 @@ func parsePairStorageNew(opts []Pair) (pairStorageNew, error) {
 			}
 			result.HasWorkDir = true
 			result.WorkDir = v.Value.(string)
+		// Default pairs
+		case "default_content_type":
+			defaultPairs = true
```
Comment (Xuanwo) [inline tests/generated.go:686]: Maybe we can simplify the logic via respect to the order of pairs?

We don't need to take `default_storage_pairs` first any more, instead we always take the order.

- If user input `WithDefaultStoragePairs` and `WithDefaultStorageClass`, we will take `WithDefaultStoragePairs` first.
- If user input `WithDefaultStorageClass` and `WithDefaultStoragePairs`, we will take `WithDefaultStorageClass`.

After this change, we don't need to maintain `defaultPairs` or `enableFeatures`, we can set `HasStorageFeatures` directly.


NOTICE: this change requires an update of GSP.

Comment (JinnyYi) [reply tests/generated.go:686]: Does this mean that we do not consider the case where there is no intersection，like: 
```
oss.WithDefaultStoragePairs("Standard")
// ...
oss.WithDefaultStoragePairs(oss.DefaultStoragePairs{Write: []types.Pair{ps.WithContentType("application/text")}})
```

Comment (Xuanwo) [reply tests/generated.go:686]: Yes, we can use them concurrently, they will not affect each other.

Comment (JinnyYi) [reply tests/generated.go:686]: Maybe we need to keep these two variables `defaultPairs` and `enableFeatures` as when there is more than one default pairs(`DefaultStorageClass` and `DefaultContentType`, or `EnableVirtualDir` and `EnableVirtualLink`), we can't set `HasDefault*Pairs` or `Has*Features` to true every time, otherwise we would ignore some default values？

Comment (Xuanwo) [reply tests/generated.go:686]: `DefaultStorageClass` and `DefaultContentType` will always be appened into `DefaultStorageParis`, I think setting `HasDefaultStorageParis` is the same with `defaultPairs`.

File: cmd/definitions/tmpl/service.tmpl [hunk 3]
```diff
@@ -103,7 +103,7 @@ func setStorageSystemMetadata(s *StorageMeta, sm StorageSystemMetadata) {
 }
 
 {{- range $_, $v := .Pairs }}
-{{- if not $v.Global }}
+{{- if and (not $v.Global) (not (hasPrefix $v.Name "enable_")) }}
```
Comment (JinnyYi) [inline cmd/definitions/tmpl/service.tmpl:106]: I removed `WithEnableXxx()` for features as we can use `<type>.WithStorageFeatures(<type>.StorageFeatures{Xxx: true, })` during initialize. I'm not sure whether it's necessary.

Comment (Xuanwo) [reply cmd/definitions/tmpl/service.tmpl:106]: I think it's OK to allow users to enable a feature via `WithEableXxx()`.

File: tests/generated.go [hunk 4]
```diff
@@ -683,34 +663,21 @@ func parsePairStorageNew(opts []Pair) (pairStorageNew, error) {
...
+			result.HasDefaultStoragePairs = true
 			result.DefaultStoragePairs.Write = append(result.DefaultStoragePairs.Write, WithStorageClass(v.Value.(string)))
 		// Enable features
 		case "enable_loose_pair":
-			if result.HasStorageFeatures {
-				continue
-			}
-			enableFeatures = true
+			result.HasStorageFeatures = true
 			result.StorageFeatures.LoosePair = v.Value.(bool)
```
Comment (JinnyYi) [inline tests/generated.go:675]: I found the last one will be picked if there are multiple `enable_loose_pair`.

File: cmd/definitions/tmpl/service.tmpl (hunk 3 above)
Comment (JinnyYi) [reply cmd/definitions/tmpl/service.tmpl:106]: OK, I'll roll back this change.

File: tests/generated.go (hunk 4 above)
Comment (Xuanwo) [reply tests/generated.go:675]: How about generating `HasEnableLoose` to prevent it?

Maybe treat `enable_loose` as a normal pair can prevent those strange edge cases?

File: services/new.go [hunk 5]
```diff
@@ -167,8 +167,12 @@ func parseConnectionString(ConnStr string) (ty string, ps []types.Pair, err erro
...
 		opt := strings.SplitN(v, "=", 2)
 		if len(opt) != 2 {
-			// && or &key&, ignore
-			continue
+			if v != "" {
+				opt = append(opt, v, "true")
+			} else {
+				// && , ignore
+				continue
+			}
```
Comment (JinnyYi) [inline services/new.go:175]: It's hard-coded here.
How about not generate pairs for `enable_xxx` features and register parsing functions for `StorageFeatures` and `ServiceFeatures`? Then we can collection pairs without values for `StorageFeatures` or `ServiceFeatures` parsing?

Comment (Xuanwo) [reply services/new.go:175]: Current implementation LGTM, we will refactor this part later, so don't be worried.

File: cmd/definitions/tmpl/service.tmpl [hunk 6]
```diff
@@ -6,6 +6,20 @@ import (
...
+    {{- range $_, $dp := $v.Defaultable }}
+    {{- range $_, $sp := $pairs }}
+    {{- if and (eq $dp.Name $sp.Name) $sp.Global }}
+        {{- $needPair = true }}
+    {{- end }}
+    {{- end }}
+    {{- end }}
+    {{- end }}
+    {{- if $needPair }}
+    "github.com/beyondstorage/go-storage/v4/pairs"
```
Comment (Xuanwo) [inline cmd/definitions/tmpl/service.tmpl:21]: How about always import pairs? Like:

```
import "github.com/beyondstorage/go-storage/v4/pairs"

var _ pairs.XXXX
```

File: services/new.go (hunk 5 above)
Comment (JinnyYi) [reply services/new.go:175]: Got it.

File: cmd/definitions/tmpl/service.tmpl (hunk 3 above)
Comment (Xuanwo) [reply cmd/definitions/tmpl/service.tmpl:106]: For bool types, maybe we can generate `WithEnableXxxx()` directly? With the internal value set to `true`.

**Review by Xuanwo — COMMENTED** [review]
I have some questions and let's discuss in https://matrix.to/#/#beyondstorage@gsp-700:matrix.org

**Comment by Xuanwo** [comment]
The PR is too big, I updated steps in https://github.com/beyondstorage/go-storage/issues/704, let's split it into PRs.

**Comment by Xuanwo** [comment]
Maybe we can close this PR for now?
